### ISSUE TITLE
Static Land interop `Observable` definition is wrong

### ISSUE THREAD

**Issue by polytypic** [kefirjs/kefir#286]
Like a comment in the current definition [says](https://github.com/kefirjs/kefir/blob/master/src/interop/static-land.js#L27-L33), the definition is wrong:

```js
  // This ap strictly speaking incompatible with chain. If we derive ap from chain we get
  // different (not very useful) behavior. But spec requires that if method can be derived
  // it must have the same behavior as hand-written method. We intentionally violate the spec
  // in hope that it won't cause many troubles in practice. And in return we have more useful type.
  ap(obsFn, obsVal) {
    return combine([obsFn, obsVal], (fn, val) => fn(val))
  },
```

I can see the "good intention" behind the above definition, but [a *major* advantage of Static Land is precisely the fact that for a single type you can simply have as many different algebra module instances as needed](https://github.com/rpominov/static-land#pros):

> * We can implement many modules for one type, therefore we can have more than one instance of the same Algebra for a single type. For example, we can implement two Monoids for numbers: Addition and Multiplication.

Observables support multiple different algebras.  For example, basically every `flatMapXXX` operation induces a Monad with different semantics.  With Static Land, each of those can be supported as different algebra modules (all of them fully correct) from which the user can then pick and choose the appropriate one depending on which semantics are needed.

So, why is the above definition wrong?  Because it breaks the sequential semantics of the chain.  Let's say that you wish to perform a *sequence* of, say, database operations using the `Observable` definition.  You should be able to do that with an applicative [`sequence`](http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Traversable.html#v:sequenceA) function, but now you cannot, because the above definition executes the operations in parallel, rather than sequentially.

Of course, another major advantage of Static Land is that this issue can be worked around simply by providing the proper algebra modules outside of the library core.

**Comment by rpominov** [comment]
Yeah, I'm still not sure what is the right call here. I don't use Static or Fantasy Land in anything real, it always was experimental/academic for me. So it's hard to tell what is the best solution from the practical point of view.

We could expose two modules, but that would complicate things. On the other hand, the single module is not quite correct.

One of the reasons why I did a single module initially is that the sequential `ap` (based on `flatMap`) seems pretty useless in the case of streams. It would spawn a second stream for each event in the first, and third for each event in a second, and so on. Basically, you get a lot of duplicates event, and the number of duplicates will grow after each event from the first stream. I can't imagine an example where this could be useful.

**Comment by polytypic** [comment]
@rpominov

> I don't use Static or Fantasy Land in anything real, it always was experimental/academic for me.

That is entirely acceptable.  I do use Static Land in real projects that are in production.

> We could expose two modules, but that would complicate things. On the other hand, the single module is not quite correct.

Yes, having more modules is more complex in a sense.  However, sometimes there is essential complexity that must not go away.

> I can't imagine an example where this could be useful.

I'll give you a pair of interconnected examples.  First of all, according to the [laws](https://github.com/rpominov/static-land/blob/master/docs/spec.md#can-be-derived-3), the functions (written in naïve style for clarity)

```js
const sequenceA = (A, ops) =>
  A.ap(A.map(R.prepend, R.head(ops)), sequenceA(A, R.tail(ops)))
```

```js
const sequenceM = (M, ops) =>
  M.chain(
    h => M.map(R.prepend(h), sequenceM(M, R.tail(ops)),
    R.head(ops)
  )
```

are equivalent (with respect to sequencing of effects) for every Monad `M`:

```js
sequenceA(M, ops) === sequenceM(M, ops)
```

An application of the laws that relate algebraic structure to each other is the ability to refactor expressions between various equivalent forms.  Of the above two definitions for sequencing, the former is preferable, because it is more polymorphic.  It works not just for every applicative, but for every monad as well, while the latter version only works for monads.

Now, let's think of a more concrete example: simple sequential [asynchronous programming](https://calmm-js.github.io/documentation/observables-are-elephants/#/2).  As hinted in those slides, observables can be used for such programming.  The gist is that each observable in such a "mode of use of observables" emits at most one value.  (This way the duplication of events does not happen.)  However, if you'd use the law-breaking `Observable` instance with `sequenceA` to sequence an array of asynchronous effects, you'd get (the wrong) parallel behavior instead of the desired sequential behavior.

**Comment by polytypic** [comment]
@hallettj

> I'm not sure I agree that ap should have sequential semantics.

Note that my point isn't that `ap` should have sequential semantics&mdash;quite the contrary actually.  My point is that in the definition of a monad, the semantics of `ap` should agree with the semantics of `chain` [as specified in the Static Land spec](https://github.com/rpominov/static-land/blob/master/docs/spec.md#can-be-derived-3).  The Static Land specification is based on corresponding specifications of monads and applicatives found in other libraries, languages, and literature.

So, instead of having one broken `Observable` module, I'm suggesting that there should rather be multiple correct modules corresponding to different ways to compose (or different modes of use of) observables.  And, indeed, there are many fundamentally different ways to use observables.

> My interpretation of what I have read is that there is no absolute rule about the relationship between ap and chain.

As mentioned above, the Static Land spec gives [such a rule](https://github.com/rpominov/static-land/blob/master/docs/spec.md#can-be-derived-3).

One basis for the rule is simply that every monad gives rise to an applicative via the definition given e.g. in [Joseph Abrahamson's SO answer](https://stackoverflow.com/questions/23342184/difference-between-monad-and-applicative-in-haskell):

```haskell
ap :: Monad m => m (a -> b) -> m a -> m b
ap mf ma = do 
  f <- mf
  a <- ma
  return (f a)
```

As also discussed in Abrahamson's answer, there are cases where an applicative instance cannot have a corresponding monad instance.

Another basis is given in [a discussions on sequencing of effects](https://en.wikibooks.org/wiki/Haskell/Applicative_functors#Sequencing_of_effects), which boils down to:

> The convention in Haskell is to always implement (<*>) and other applicative operators using left-to-right sequencing.
